### DISCUSSION TITLE
How to retrieve a function's parameter list (function prototype) using Python in Ghidra?

### DISCUSSION THREAD

**Discussion by WordDealer** [NationalSecurityAgency/ghidra#5852] [Q&A]
I am trying to retrieve the parameter list for a function named rpl_fseeko in a binary file. My aim is to determine how many parameters this function has. However, I am uncertain about the correct way to use the API for this purpose. Here's what I've tried:
```
func_manager = currentProgram.getFunctionManager()
func_iterator = func_manager.getFunctions(True)
for func in func_iterator:
    print(func.getName())
    print(func.getAllVariables())
    if func.getName() == "rpl_fseeko":
        print(type(func))
        print(dir(func))
        print(func.parameters)
        for x in func.parameter:
            print(x)
        print(len(func.getParameters()))
        exit()

```
This always returns 0 parameters. However, I can observe 3 parameters in the decompiled code in Ghidra. How should I correctly use the API to get the expected result of 3 parameters?

**Comment by anzosasuke** [comment]
you code should be working. Just one slight error in the code:
instead of this:
``for x in func.parameter:
            print(x)
``
do this:

``param = function.getParameters()
  for i in param:
      print(i)
``

I don't think there is an API parameters. Look in ghidra.re and go through these API.
[https://ghidra.re/ghidra_docs/api/ghidra/program/model/listing/Function.html#getParameter(int)](url)

**Reply by WordDealer** [reply]
This is F5 in IDA Pro,
![image](https://github.com/NationalSecurityAgency/ghidra/assets/56788639/044d33c4-df12-4d9c-9e3f-80f90cd357d4)
this function has 3 parameters, but function.getParameters() will get 0 .
and this is "F5" in Ghidra:
![image](https://github.com/NationalSecurityAgency/ghidra/assets/56788639/a2c107c4-9602-41b3-8462-94ce636c6ce8)
I don not know why?maybe I need to let ghidra analyze `Function calling convention` first, so how can I get parameters list like F5's answer

**Reply by WordDealer** [reply]
Both two tools of 'F5' is right, but I want to use Ghidra python api to get these 3 parameters . and the code in my first question get 0 parameter

**Reply by anzosasuke** [reply]
Im no expert but post the binary so others can help and recreate it as well.

**Reply by WordDealer** [reply]
"Thank you, I have posted the binary file and Python script.

**Comment by WordDealer** [comment]
This is the result in Ghidra:
![image](https://github.com/NationalSecurityAgency/ghidra/assets/56788639/f91f9b7a-daca-48fc-a3d7-9e9227d437a7)
This is binary file(from coreutils) and test.py
[test.zip](https://github.com/NationalSecurityAgency/ghidra/files/12816495/test.zip)
And the right answer is 3, not 0.

**Reply by anzosasuke** [reply]
You are right, we can't get it from disassembly. However, you can get it from decompiler. Probably for ghidra to get parameters from disassembly, it may look for RDI,RSI or other registers which in this case is not that evident.
```

from ghidra.program.model.listing import FunctionSignature
func_manager=currentProgram.getFunctionManager()
func_iterator=func_manager.getFunctions(True)
from ghidra.program.model.pcode import FunctionPrototype

def get_high_function(func):
    options = DecompileOptions()
    ifc = DecompInterface()
    ifc.setOptions(options)
    ifc.openProgram(getCurrentProgram())
    res = ifc.decompileFunction(func, 60, monitor)
    high = res.getHighFunction()
    return high


for func in func_iterator:

    if func.getName()=="rpl_fseeko":
        hf = get_high_function(func)
        print(hf.getFunctionPrototype().getNumParams())
        for i in range(hf.getFunctionPrototype().getNumParams()):
            print(hf.getFunctionPrototype().getParam(i).getName())

```

**Reply by graps01** [reply]
Newbie question: On a command line in debian, would I invoke the python code first, then analyze it in ghidra ? Or, run the cp executable, then save the executable's output to a file for analysis ?

**Reply by astrelsky** [reply]
> You are right, we can't get it from disassembly. However, you can get it from decompiler. Probably for ghidra to get parameters from disassembly, it may look for RDI,RSI or other registers which in this case is not that evident.
> ```
> 
> from ghidra.program.model.listing import FunctionSignature
> func_manager=currentProgram.getFunctionManager()
> func_iterator=func_manager.getFunctions(True)
> from ghidra.program.model.pcode import FunctionPrototype
> 
> def get_high_function(func):
>     options = DecompileOptions()
>     ifc = DecompInterface()
>     ifc.setOptions(options)
>     ifc.openProgram(getCurrentProgram())
>     res = ifc.decompileFunction(func, 60, monitor)
>     high = res.getHighFunction()
>     return high
> 
> 
> for func in func_iterator:
> 
>     if func.getName()=="rpl_fseeko":
>         hf = get_high_function(func)
>         print(hf.getFunctionPrototype().getNumParams())
>         for i in range(hf.getFunctionPrototype().getNumParams()):
>             print(hf.getFunctionPrototype().getParam(i).getName())
> 
> ```
> 

Obtaining the parameters from the listing Function object as seen in the initial post returns the committed parameters (the ones explicitly applied by the user or by applying a function signature data type).

The example code @anzosasuke shows above is providing the parameters that the decompiler has determined them to be based on its analysis. These parameters have not been committed and will change as further analysis is performed on other functions and other functions parameters are committed.
